cert/ca.go

```diff
@@ -0,0 +1,166 @@
+// Package cert contains certificate specifications and
+// certificate-specific management.
+package cert
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/cloudflare/cfssl/api/client"
+	"github.com/cloudflare/cfssl/info"
+	"github.com/cloudflare/cfssl/log"
+)
+
+// A CA contains the core details for a CFSSL CA. There are two ways
+// to use this: fill out Name to refer to a global CA (e.g. as defined
+// in the config file) or fill out Remote, Label, Profile, and AuthKey.
+type CA struct {
+	Name        string `json:"name" yaml:"name"`
+	Remote      string `json:"remote" yaml:"remote"`
+	Label       string `json:"label" yaml:"label"`
+	Profile     string `json:"profile" yaml:"profile"`
+	AuthKey     string `json:"auth_key" yaml:"auth_key"`
+	AuthKeyFile string `json:"auth_key_file" yaml:"auth_key_file"`
+	File        *File  `json:"file,omitempty" yaml:"file,omitempty"`
+	RootCACert  string `json:"root_ca,omitempty" yaml:"root_ca,omitempty"`
+	pem         []byte
+}
+
+// GetPEM is for testing only!
+// Getter for CA cert PEM
+func (ca *CA) GetPEM() []byte {
+	return ca.pem
+}
+
+// SetPEM is for testing only!
+// Setter for CA cert PEM
+func (ca *CA) SetPEM(pem []byte) {
+	ca.pem = pem
+}
+
+func (ca *CA) getRemoteCert() ([]byte, error) {
+	var tlsConfig *tls.Config
+	if ca.RootCACert != "" {
+		rootCABytes, err := ioutil.ReadFile(ca.RootCACert)
+		if err != nil {
+			return nil, err
+		}
+
+		rootCaCertPool := x509.NewCertPool()
+		ok := rootCaCertPool.AppendCertsFromPEM(rootCABytes)
+		if !ok {
+			return nil, errors.New("failed to parse rootCA certs")
+		}
+		tlsConfig = &tls.Config{
+			RootCAs: rootCaCertPool,
+		}
+	}
+
+	remote := client.NewServerTLS(ca.Remote, tlsConfig)
+	infoReq := &info.Req{
+		Label:   ca.Label,
+		Profile: ca.Profile,
+	}
+
+	serialisedRequest, err := json.Marshal(infoReq)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := remote.Info(serialisedRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(strings.TrimSpace(resp.Certificate)), nil
```

Comment:
I'm 99% sure the trim space here isn't necessary anymore. That's on me though -- I should've changed this earlier.

Reply:
My intent actually is to just leverage the hell out of that x509 comparison logic you added, and drop all of this- instead just defer to doing comparisons via that.  I was intending on doing the same thing for the mtime checks- we don't care if the spec is newer then cert/key on disk as long as x509 and key material is the same.  Which removes yet another errant restart....

Until then, I'd prefer we leave the trim alone- the code's going to die, and that pathway is a known source of incorrect restarts.